File: dashboards/dashboards/page/cash_dashboard/cash_dashboard.js
frappe.pages["cash-dashboard"].on_page_load = function (wrapper) {
	new dashboards.ui.CashDashboardPage(wrapper);
};

frappe.provide("dashboards.ui");

dashboards.ui.CashDashboardPage = class CashDashboardPage {
	constructor(wrapper) {
		this.wrapper = $(wrapper);
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Cash Dashboard"),
			single_column: true,
		});

		this.state = {
			month: null,
		};

		this.make_layout();
		this.load_context();
	}

	make_layout() {
		this.wrapper.find(".layout-main-section-wrapper").addClass("cash-dashboard-layout");
		this.wrapper.find(".page-head").addClass("cash-dashboard-page-head");
		this.page.main.removeClass("frappe-card");

		this.page.main.html(`
			<div class="cash-dashboard-screen">
				<div class="cash-dashboard-shell">
					<div class="cash-dashboard-tabs" data-region="tabs"></div>
					<div class="cash-dashboard-months" data-region="months"></div>
					<div class="cash-dashboard-kpis">
						<div class="cash-dashboard-kpi cash-dashboard-kpi--wide" data-region="cash-kpi"></div>
						<div class="cash-dashboard-kpi cash-dashboard-kpi--balance" data-region="cash-balance"></div>
						<div class="cash-dashboard-kpi cash-dashboard-kpi--balance" data-region="bank-balance"></div>
						<div class="cash-dashboard-kpi cash-dashboard-kpi--wide" data-region="bank-kpi"></div>
					</div>
					<div class="cash-dashboard-content">
						<div class="cash-dashboard-table-slot" data-region="cash-table"></div>
						<div class="cash-dashboard-table-slot" data-region="bank-table"></div>
					</div>
				</div>
			</div>
		`);

		this.$tabs = this.page.main.find('[data-region="tabs"]');
		this.$months = this.page.main.find('[data-region="months"]');
		this.$cashKpi = this.page.main.find('[data-region="cash-kpi"]');
		this.$cashBalance = this.page.main.find('[data-region="cash-balance"]');
		this.$bankBalance = this.page.main.find('[data-region="bank-balance"]');
		this.$bankKpi = this.page.main.find('[data-region="bank-kpi"]');
		this.$cashTable = this.page.main.find('[data-region="cash-table"]');
		this.$bankTable = this.page.main.find('[data-region="bank-table"]');
	}

	load_context() {
		frappe.call({
			method: "dashboards.dashboards.page.cash_dashboard.cash_dashboard.get_dashboard_context",
			callback: (r) => {
				this.context = r.message || {};
				this.state = { ...(this.context.default_filters || {}) };
				this.render();
			},
		});
	}

	render() {
		this.render_tabs();
		this.render_months();
		this.render_kpis();
		this.render_tables();
	}

	render_tabs() {
		const tabs = this.context.tabs || [];
		this.$tabs.html(
			tabs
				.map(
					(tab) => `
						<button class="cash-dashboard-tab ${tab.active ? "is-active" : ""}" data-route="${tab.route}">
							${frappe.utils.escape_html(tab.label)}
						</button>
					`
				)
				.join("") +
				`<div class="cash-dashboard-info">i</div>`
		);

		this.$tabs.find("[data-route]").on("click", (e) => {
			const route = $(e.currentTarget).data("route");
			if (route) {
				frappe.set_route(route.replace(/^\/app\//, ""));
			}
		});
	}

	render_months() {
		const months = this.context.months || [];
		this.$months.html(
			months
				.map(
					(month) => `
						<button class="cash-dashboard-month ${month.key === this.state.month ? "is-active" : ""}" data-month="${month.key}">
							${frappe.utils.escape_html(month.label)}
						</button>
					`
				)
				.join("")
		);

		this.$months.find("[data-month]").on("click", (e) => {
			this.state.month = String($(e.currentTarget).data("month"));
			this.render_kpis();
			this.render_tables();
			this.render_months();
		});
	}

	render_kpis() {
		const metrics = this.getMetrics();

		this.$cashKpi.html(this.getWideKpiMarkup(metrics.cash, "касса"));
		this.$bankKpi.html(this.getWideKpiMarkup(metrics.bank, "банк"));
		this.$cashBalance.html(this.getBalanceMarkup(metrics.cash.end, "Остаток касса"));
		this.$bankBalance.html(this.getBalanceMarkup(metrics.bank.end, "Остаток банк"));
	}

	render_tables() {
		const metrics = this.getMetrics();
		const cashRows = this.getRowsForMonth(this.context.cash_rows || [], metrics.cash.inflow, "cash");
		const bankRows = this.getRowsForMonth(this.context.bank_rows || [], metrics.bank.inflow, "bank");
		this.$cashTable.html(this.getTableMarkup(cashRows, metrics.cash, "касса"));
		this.$bankTable.html(this.getTableMarkup(bankRows, metrics.bank, "банк"));
	}

	getWideKpiMarkup(metric, label) {
		return `
			<div class="cash-dashboard-kpi-grid">
				<div class="cash-dashboard-kpi-item">
					<div class="cash-dashboard-kpi-value is-green">${this.formatInteger(metric.start)}</div>
					<div class="cash-dashboard-kpi-label">Начало ${label}</div>
				</div>
				<div class="cash-dashboard-kpi-item">
					<div class="cash-dashboard-kpi-value is-green">${this.formatInteger(metric.inflow)}</div>
					<div class="cash-dashboard-kpi-label">Приход ${label}</div>
				</div>
				<div class="cash-dashboard-kpi-item">
					<div class="cash-dashboard-kpi-value is-red">${this.formatInteger(metric.outflow)}</div>
					<div class="cash-dashboard-kpi-label">Расход ${label}</div>
				</div>
				<div class="cash-dashboard-kpi-item">
					<div class="cash-dashboard-kpi-value is-green">${this.formatInteger(metric.end)}</div>
					<div class="cash-dashboard-kpi-label">Конец ${label}</div>
				</div>
			</div>
		`;
	}

	getBalanceMarkup(value, label) {
		return `
			<div class="cash-dashboard-balance-value">${this.formatBalance(value)}</div>
			<div class="cash-dashboard-balance-label">${frappe.utils.escape_html(label)}</div>
		`;
	}

	getTableMarkup(rows, metric, label) {
		return `
			<table class="cash-dashboard-table">
				<thead>
					<tr>
						<th class="is-category">(2)</th>
						<th class="is-number">Прих.${frappe.utils.escape_html(label)}</th>
						<th class="is-number">Расход ${frappe.utils.escape_html(label)}</th>
					</tr>
				</thead>
				<tbody>
					${rows
						.map(
							(row) => `
								<tr class="${row.group ? "is-group" : "is-child"}">
									<td class="is-category ${row.level ? "is-level-1" : ""}">
										${row.group ? '<span class="cash-dashboard-tree-toggle">⊞</span>' : ""}
										<span>${frappe.utils.escape_html(row.label)}</span>
									</td>
									<td class="is-number">${row.inflow ? this.formatInteger(row.inflow) : ""}</td>
									<td class="is-number">${row.outflow ? this.formatInteger(row.outflow) : "0"}</td>
								</tr>
							`
						)
						.join("")}
					<tr class="is-total">
						<td class="is-category">Total</td>
						<td class="is-number">${this.formatInteger(metric.inflow)}</td>
						<td class="is-number">${this.formatInteger(metric.outflow)}</td>
					</tr>
				</tbody>
			</table>
		`;
	}

	getMetrics() {
		const monthIndex = (this.context.months || []).findIndex((month) => month.key === this.state.month);
		const monthFactor = 0.72 + (Math.max(monthIndex, 0) * 0.04);
		const cashBase = this.context.base_kpi?.cash || {};
		const bankBase = this.context.base_kpi?.bank || {};

		const cash = {
			start: Math.round((cashBase.start || 0) * (0.94 + monthFactor * 0.08)),
			target_inflow: Math.round((cashBase.inflow || 0) * monthFactor),
			target_outflow: Math.round((cashBase.outflow || 0) * (0.71 + Math.max(monthIndex, 0) * 0.039)),
		};
		const cashRows = this.getRowsForMonth(this.context.cash_rows || [], cash.target_inflow, "cash");
		cash.inflow = this.sumField(cashRows, "inflow");
		cash.outflow = this.sumField(cashRows, "outflow");
		cash.end = cash.start + cash.inflow - cash.outflow;

		const bank = {
			start: Math.round((bankBase.start || 0) * (0.82 + monthFactor * 0.05)),
			target_inflow: Math.round((bankBase.inflow || 0) * monthFactor),
			target_outflow: Math.round((bankBase.outflow || 0) * (0.74 + Math.max(monthIndex, 0) * 0.033)),
		};
		const bankRows = this.getRowsForMonth(this.context.bank_rows || [], bank.target_inflow, "bank");
		bank.inflow = this.sumField(bankRows, "inflow");
		bank.outflow = this.sumField(bankRows, "outflow");
		bank.end = bank.start + bank.inflow - bank.outflow;

		return { cash, bank };
	}

	getRowsForMonth(rows, salesInflowTarget, key) {
		const monthIndex = (this.context.months || []).findIndex((month) => month.key === this.state.month);
		const factor = 0.74 + (Math.max(monthIndex, 0) * 0.037);

		return rows.map((row, index) => {
			const variance = 1 + ((index % 3) * 0.016);
			let inflow = Math.round((row.inflow || 0) * factor * variance);
			let outflow = Math.round((row.outflow || 0) * factor * (0.99 + ((monthIndex + index) % 2) * 0.02));

			if (key === "cash" && row.label === "Продажа") {
				inflow = salesInflowTarget;
			}

			if (key === "bank" && row.label === "Продажа") {
				inflow = salesInflowTarget;
			}

			return {
				...row,
				inflow,
				outflow,
			};
		});
	}

	sumField(rows, fieldname) {
		return (rows || []).reduce((total, row) => total + (row[fieldname] || 0), 0);
	}

	formatInteger(value) {
		const sign = value < 0 ? "-" : "";
		const numeric = Math.abs(Math.round(value));
		return `${sign}${String(numeric).replace(/\B(?=(\d{3})+(?!\d))/g, " ")}`;
	}

	formatBalance(value) {
		if (!value) {
			return "(Blank)";
		}

		const millions = Math.round(value / 1000000);
		return `${millions}M`;
	}
};

File: dashboards/dashboards/page/cash_dashboard/cash_dashboard.py
from __future__ import annotations

from typing import Any

import frappe
from frappe.utils import flt, getdate, get_first_day, get_last_day, today

from dashboards.dashboards.dashboard_data import MONTH_LABELS


TAB_ITEMS = [
	{"label": "ГЛАВНЫЙ", "route": "/app/main-dashboard"},
	{"label": "ПАНЕЛЬ", "route": "/app/page-dashboard"},
	{"label": "ЕЖЕДНЕВНО", "route": "/app/daily-dashboard"},
	{"label": "ПРОДАЖА", "route": "/app/sales-dashboard"},
	{"label": "КАССА", "route": "/app/cash-dashboard", "active": 1},
	{"label": "КЛИЕНТ", "route": "/app/client-dashboard"},
	{"label": "ПОСТАВЩИКИ", "route": "/app/supplier-dashboard"},
]

MONTHS = [{"key": label.lower(), "label": label} for label in MONTH_LABELS]
MONTH_MAP = {item["key"]: index + 1 for index, item in enumerate(MONTHS)}
ACCOUNT_TYPE_LABELS = {"Cash": "касса", "Bank": "банк"}


def _get_reference_date():
	row = frappe.db.sql(
		"""
		SELECT MAX(posting_date) AS posting_date
		FROM `tabGL Entry`
		WHERE is_cancelled = 0
		""",
		as_dict=True,
	)[0]
	return getdate(row.posting_date) if row.posting_date else getdate(today())


def _get_default_period() -> tuple[str, str]:
	reference_date = _get_reference_date()
	return str(reference_date.year), MONTH_LABELS[reference_date.month - 1].lower()


def _normalize_filters(month: str | None) -> tuple[str, str]:
	default_year, default_month = _get_default_period()
	selected_month = month if month in MONTH_MAP else default_month
	return default_year, selected_month


def _get_period_range(year: str, month: str) -> tuple[str, str]:
	month_no = MONTH_MAP[month]
	reference_date = getdate(f"{year}-{month_no:02d}-01")
	return str(get_first_day(reference_date)), str(get_last_day(reference_date))


def _get_balance(account_type: str, before_date: str | None = None, end_date: str | None = None) -> float:
	clauses = []
	params: dict[str, Any] = {"account_type": account_type}

	if before_date:
		clauses.append("AND gle.posting_date < %(before_date)s")
		params["before_date"] = before_date

	if end_date:
		clauses.append("AND gle.posting_date <= %(end_date)s")
		params["end_date"] = end_date

	row = frappe.db.sql(
		f"""
		SELECT
			SUM(COALESCE(gle.debit, 0) - COALESCE(gle.credit, 0)) AS balance
		FROM `tabGL Entry` gle
		INNER JOIN `tabAccount` acc ON acc.name = gle.account
		WHERE gle.is_cancelled = 0
		  AND acc.is_group = 0
		  AND acc.account_type = %(account_type)s
		  {' '.join(clauses)}
		""",
		params,
		as_dict=True,
	)[0]
	return flt(row.balance)


def _get_flow_rows(account_type: str, start_date: str, end_date: str) -> list[dict[str, Any]]:
	rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(NULLIF(gle.voucher_type, ''), 'Другое') AS label,
			SUM(COALESCE(gle.debit, 0)) AS inflow,
			SUM(COALESCE(gle.credit, 0)) AS outflow
		FROM `tabGL Entry` gle
		INNER JOIN `tabAccount` acc ON acc.name = gle.account
		WHERE gle.is_cancelled = 0
		  AND acc.is_group = 0
		  AND acc.account_type = %(account_type)s
		  AND gle.posting_date BETWEEN %(start_date)s AND %(end_date)s
		GROUP BY COALESCE(NULLIF(gle.voucher_type, ''), 'Другое')
		ORDER BY (SUM(COALESCE(gle.debit, 0)) + SUM(COALESCE(gle.credit, 0))) DESC, label ASC
		""",
		{
			"account_type": account_type,
			"start_date": start_date,
			"end_date": end_date,
		},
		as_dict=True,
	)

	result = []
	for row in rows:
		result.append(
			{
				"label": row.label,
				"inflow": round(flt(row.inflow)),
				"outflow": round(flt(row.outflow)),
				"level": 0,
				"group": True,
			}
		)

	return result


def _build_section(account_type: str, start_date: str, end_date: str) -> dict[str, Any]:
	rows = _get_flow_rows(account_type, start_date, end_date)
	inflow = sum(flt(row["inflow"]) for row in rows)
	outflow = sum(flt(row["outflow"]) for row in rows)
	start_balance = _get_balance(account_type, before_date=start_date)
	end_balance = start_balance + inflow - outflow

	return {
		"label": ACCOUNT_TYPE_LABELS[account_type],
		"rows": rows,
		"metric": {
			"start": round(start_balance),
			"inflow": round(inflow),
			"outflow": round(outflow),
			"end": round(end_balance),
		},
	}


@frappe.whitelist()
def get_dashboard_context(month: str | None = None):
	selected_year, selected_month = _normalize_filters(month)
	start_date, end_date = _get_period_range(selected_year, selected_month)
	cash = _build_section("Cash", start_date, end_date)
	bank = _build_section("Bank", start_date, end_date)

	return {
		"tabs": TAB_ITEMS,
		"default_filters": {"month": selected_month},
		"months": MONTHS,
		"period_label": f"{MONTH_LABELS[MONTH_MAP[selected_month] - 1]} {selected_year}",
		"cash_rows": cash["rows"],
		"bank_rows": bank["rows"],
		"metrics": {
			"cash": cash["metric"],
			"bank": bank["metric"],
		},
	}
